wp-includes/class-wp-service-workers.php

```diff
@@ -0,0 +1,216 @@
+			if ( $scope === $item->args ) {
+				$scope_items[] = $handle;
+			}
+		}
+
+		$this->output = '';
+		$this->do_items( $scope_items );
+
+		$file_hash = md5( $this->output );
+		header( "Etag: $file_hash" );
+
+		$etag_header = isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ? trim( $_SERVER['HTTP_IF_NONE_MATCH'] ) : false;
+		if ( $file_hash === $etag_header ) {
+			status_header( 304 );
+			return;
+		}
+
+		// @codingStandardsIgnoreLine
+		echo $this->output;
+	}
+
+	/**
+	 * Get all scopes.
+	 *
+	 * @return array Array of scopes.
+	 */
+	public function get_scopes() {
+
+		$scopes = array();
+		foreach ( $this->registered as $handle => $item ) {
+			$scopes[] = $item->args;
+		}
+		return $scopes;
+	}
+
+	/**
+	 * Process one registered script.
+	 *
+	 * @param string $handle Handle.
+	 * @param bool   $group Group.
+	 * @return void
+	 */
+	public function do_item( $handle, $group = false ) {
+		global $wp_filesystem;
+
+		$obj           = $this->registered[ $handle ];
+		$this->output .= $wp_filesystem->get_contents( site_url() . $obj->src ) . "\n";
+	}
+
+	/**
+	 * Remove URL scheme.
+	 *
+	 * @param string $schemed_url URL.
+	 * @return string URL.
+	 */
+	public function remove_url_scheme( $schemed_url ) {
+		return preg_replace( '#^\w+:(?=//)#', '', $schemed_url );
+	}
+
+	/**
+	 * Get validated path to file.
+	 *
+	 * @param string $url Relative path.
+	 * @return null|string|WP_Error
+	 */
+	public function get_validated_file_path( $url ) {
+		$needs_base_url = ! preg_match( '|^(https?:)?//|', $url );
+		$base_url       = site_url();
+
+		if ( $needs_base_url ) {
+			$url = $base_url . $url;
+		}
+
+		// Strip URL scheme, query, and fragment.
+		$url = $this->remove_url_scheme( preg_replace( ':[\?#].*$:', '', $url ) );
+
+		$includes_url = $this->remove_url_scheme( includes_url( '/' ) );
+		$content_url  = $this->remove_url_scheme( content_url( '/' ) );
+		$admin_url    = $this->remove_url_scheme( get_admin_url( null, '/' ) );
+
+		$allowed_hosts = array(
+			wp_parse_url( $includes_url, PHP_URL_HOST ),
+			wp_parse_url( $content_url, PHP_URL_HOST ),
+			wp_parse_url( $admin_url, PHP_URL_HOST ),
+		);
+
+		$url_host = wp_parse_url( $url, PHP_URL_HOST );
+
+		if ( ! in_array( $url_host, $allowed_hosts, true ) ) {
+			/* translators: %s is file URL */
+			return new WP_Error( 'external_file_url', sprintf( __( 'URL is located on an external domain: %s.', 'pwa' ), $url_host ) );
+		}
+
+		$file_path = null;
+		if ( 0 === strpos( $url, $content_url ) ) {
+			$file_path = WP_CONTENT_DIR . substr( $url, strlen( $content_url ) - 1 );
+		} elseif ( 0 === strpos( $url, $includes_url ) ) {
+			$file_path = ABSPATH . WPINC . substr( $url, strlen( $includes_url ) - 1 );
+		} elseif ( 0 === strpos( $url, $admin_url ) ) {
+			$file_path = ABSPATH . 'wp-admin' . substr( $url, strlen( $admin_url ) - 1 );
+		} else {
+			$file_path = ABSPATH . substr( $url, strlen( $this->remove_url_scheme( $base_url ) ) );
```

Comment:
Also added the `else` here for now.

Wondering if we should restrict the SW file path to validating against `WP_CONTENT_DIR` only? Then the potential default core WP SW scripts could be registered via `WP_Service_Workers::register_default_service_workers` which would skip the check for `get_validated_file_path`. Not sure if plugins/themes should be able to register scripts that are outside of `WP_CONTENT_DIR`. Thoughts?

Reply:
Since plugins and themes would only enqueue their scripts that are inside `WP_CONTENT_DIR` I think it is safe to assume that the same would be true for service worker scripts.